core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -180,131 +266,55 @@ public static Script parse(XContentParser parser, ParseFieldMatcher matcher, Str
-                        "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<options>]");
-                }
-            } else if (matcher.match(name, PARAMS_PARSE_FIELD)) {
-                if (token == Token.START_OBJECT) {
-                    params = parser.map();
-                } else {
-                    throw new ParsingException(parser.getTokenLocation(),
-                        "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<params>]");
-                }
-            } else {
-                throw new ParsingException(parser.getTokenLocation(),
-                    "unexpected field [" + (name == null ? parser.currentToken() : name) + "], " + "expected [" +
-                        ScriptType.INLINE.getParseField().getPreferredName() + ", " +
-                        ScriptType.STORED.getParseField().getPreferredName() + ", " +
-                        ScriptType.FILE.getParseField().getPreferredName() + ", " +
-                        LANG_PARSE_FIELD.getPreferredName() + ", " +
-                        PARAMS_PARSE_FIELD.getPreferredName() +
-                        "]");
-            }
+        private ParsedCode(String idOrCode, String contentType) {
+            this.idOrCode = idOrCode;
+            this.contentType = contentType;
         }
+    }
 
-        if (type == null) {
-            throw new IllegalArgumentException("missing required parameter for parsing script, expected valid script type [" +
-                ScriptType.INLINE.getName() + "] or [" + ScriptType.STORED.getName() + "] or [" + ScriptType.FILE.getName() + "]");
+    /**
+     * Parses either a simple {@link String} for {@link Script#idOrCode} or converts a JSON object into a {@link String}
+     * also saving off the appropriate {@link XContentType} to be used for compilation if necessary.
+     */
+    private static ParsedCode parseCode(XContentParser parser, ParseFieldMatcherSupplier supplier) throws IOException {
+        String idOrCode;
+        String contentType = null;
+
+        if (parser.currentToken() == Token.START_OBJECT) {
+            XContentBuilder builder = XContentFactory.contentBuilder(parser.contentType());
+            idOrCode = builder.copyCurrentStructure(parser).bytes().utf8ToString();
+            contentType = parser.contentType().mediaType();
+        } else {
+            idOrCode = parser.text();
         }
 
-        if (idOrCode == null) {
-            throw new IllegalArgumentException("missing required parameter for parsing script, expected [<id>] or [<code>]");
-        }
+        return new ParsedCode(idOrCode, contentType);
+    }
 
-        if (params == null) {
-            params = new HashMap<>();
-        }
+    /**
+     * Parses compiler options.
+     */
+    private static Map<String, String> parseOptions(XContentParser parser, ParseFieldMatcherSupplier supplier) throws IOException {
```

Comment:
I *think* you can do `XContentParser::mapStrings` above instead of having this method.